docs/src/main/sphinx/sql/pattern-recognition-in-window.rst

```diff
@@ -0,0 +1,249 @@
+============================================
+
+A window structure can be defined in the ``WINDOW`` clause or in the ``OVER``
+clause of a window operation. In both cases, the window specification can
+include row pattern recognition clauses. They are part of the window frame. The
+syntax and semantics of row pattern recognition in window are similar to those
+of the :doc:`MATCH_RECOGNIZE</sql/match-recognize>` clause.
+
+This section explains the details of row pattern recognition in window
+structures, and highlights the similarities and the differences between both
+pattern recognition mechanisms.
+
+Window with row pattern recognition
+-----------------------------------
+
+**Window specification:**
+
+.. code-block:: text
+
+        (
+        [ existing_window_name ]
+        [ PARTITION BY column [, ...] ]
+        [ ORDER BY column [, ...] ]
+        [ window_frame ]
+        )
+
+**Window frame:**
+
+.. code-block:: text
+
+        [ MEASURES measure_definition [, ...] ]
+        frame_extent
+        [ AFTER MATCH skip_to ]
+        [ INITIAL | SEEK ]
+        [ PATTERN ( row_pattern ) ]
+        [ SUBSET subset_definition [, ...] ]
+        [ DEFINE variable_definition [, ...] ]
+
+Generally, a window frame specifies the ``frame_extent``, which defines the
+"sliding window" of rows to be processed by a window function. It can be
+defined in terms of ``ROWS``, ``RANGE`` or ``GROUPS``.
+
+A window frame with row pattern recognition involves many other syntactical
+components, mandatory or optional, and enforces certain limitations on the
+``frame_extent``.
+
+**Window frame with row pattern recognition:**
+
+.. code-block:: text
+
+        [ MEASURES measure_definition [, ...] ]
+        ROWS BETWEEN CURRENT ROW AND frame_end
+        [ AFTER MATCH skip_to ]
+        [ INITIAL | SEEK ]
+        PATTERN ( row_pattern )
+        [ SUBSET subset_definition [, ...] ]
+        DEFINE variable_definition [, ...]
+
+Description of the pattern recognition clauses
+----------------------------------------------
+
+The ``frame_extent`` with row pattern recognition must be defined in terms of
+``ROWS``. The frame start must be at the ``CURRENT ROW``, which limits the
+allowed frame extent values to the following::
+
+    ROWS BETWEEN CURRENT ROW AND CURRENT ROW
+
+    ROWS BETWEEN CURRENT ROW AND <expression> FOLLOWING
+
+    ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING
+
+For every input row processed by the window, the portion of rows enclosed by
+the ``frame_extent`` limits the search area for row pattern recognition. Unlike
+in ``MATCH_RECOGNIZE``, where the pattern search can explore all rows until the
+partition end, and all rows of the partition are available for computations, in
+window structures the pattern matching can neither match rows nor retrieve
+input values outside the frame.
+
+Besides the ``frame_extent``, the two required clauses for pattern matching are
+``PATTERN`` and ``DEFINE``.
```

Comment:
Besides the ``frame_extent``, pattern matching requires the ``PATTERN`` and ``DEFINE`` clauses.